### PR TITLE
improve gru unit performance.

### PR DESCRIPTION
**Pull request by zhaoyuchen2018**
test=develop
original performance:
image
after opt:
image

Signed-off-by: zhaoyuchen <zhaoyuchen01@baidu.com>

### CODE REVIEW

**Comment by Xreki** [comment]
I test it in `1d-attention` model, the inference time is reduced from `8.12913ms` to `7.82712ms`. The detailed profiling result:
- Before
```
Event                                         Calls       Total       CPU Time (Ratio)        GPU Time (Ratio)        Min.        Max.        Ave.        Ratio.  
thread0::gru                                  200         336.763     254.454392 (0.755590)   82.308119 (0.244410)    1.66599     1.73251     1.68381     0.102072
```
- After
```
Event                                         Calls       Total       CPU Time (Ratio)        GPU Time (Ratio)        Min.        Max.        Ave.        Ratio.
thread0::gru                                  200         300.651     150.755833 (0.501431)   149.895415 (0.498569)   1.40627     7.8125      1.50326     0.093955
```

So why the CPU time is reduced and the GPU time is increased? 

I noticed there are `CPU->GPU` memory, is it copying `lod`?
@qingqing01 would you please help to review?

And there is a `gru_unit op`, @zhaoyuchen2018 would you please optimize it by the way?

**Comment by zhaoyuchen2018** [comment]
> I test it in `1d-attention` model, the inference time is reduced from `8.12913ms` to `7.82712ms`. The detailed profiling result:
> 
> * Before
> 
> ```
> Event                                         Calls       Total       CPU Time (Ratio)        GPU Time (Ratio)        Min.        Max.        Ave.        Ratio.  
> thread0::gru                                  200         336.763     254.454392 (0.755590)   82.308119 (0.244410)    1.66599     1.73251     1.68381     0.102072
> ```
> * After
> 
> ```
> Event                                         Calls       Total       CPU Time (Ratio)        GPU Time (Ratio)        Min.        Max.        Ave.        Ratio.
> thread0::gru                                  200         300.651     150.755833 (0.501431)   149.895415 (0.498569)   1.40627     7.8125      1.50326     0.093955
> ```
> So why the CPU time is reduced and the GPU time is increased?
> 
> I noticed there are `CPU->GPU` memory, is it copying `lod`?
> @qingqing01 would you please help to review?
> 
> And there is a `gru_unit op`, @zhaoyuchen2018 would you please optimize it by the way?

More work did in gpu， so cost much time， cpu time reduced because kernel number decreased。but if gpu ability higher，gpu time may decreased。

Will add new pr for gru unit op

File: paddle/fluid/operators/math/gru_compute.cu [hunk 1]
```diff
@@ -30,10 +30,26 @@ struct GRUUnitFunctor<platform::CUDADeviceContext, T> {
     dim3 threads;
     dim3 grid;
     if (batch_size == 1) {
-      int frame_per_block = frame_size <= 1024 ? frame_size : 1024;
-      int frame_blocks = (frame_size + 1024 - 1) / 1024;
-      threads = dim3(frame_per_block, 1);
+      const int tiled_size = 16;
```
Comment (tensor-tang) [inline paddle/fluid/operators/math/gru_compute.cu]: use `constexpr`

File: paddle/fluid/operators/math/gru_compute.cu [hunk 2]
```diff
@@ -30,10 +30,26 @@ struct GRUUnitFunctor<platform::CUDADeviceContext, T> {
...
+      int frame_blocks = (frame_size * 2 + tiled_size - 1) / tiled_size;
+      threads = dim3(tiled_size, 1);
       grid = dim3(frame_blocks, 1);
+
+      detail::KeFastCollectiveGruGate<T,
```
Comment (tensor-tang) [inline paddle/fluid/operators/math/gru_compute.cu:38]: What is the meaning of `KeFastCollective`?

File: paddle/fluid/operators/math/detail/gru_gpu_kernel.h [hunk 3]
```diff
@@ -101,6 +101,114 @@ __global__ void KeGruForwardFinalOutput(OpFinalOutput op_final_output,
   output_value[frame_idx] = r_output;
 }
 
+template <class T, int Tiled_size>
```
Comment (tensor-tang) [inline paddle/fluid/operators/math/detail/gru_gpu_kernel.h:108]: give some brief introduction like https://github.com/PaddlePaddle/Paddle/pull/16338/files#diff-49acf385fbbb7d1ed37c61818fe9de8fR212

File: paddle/fluid/operators/math/gru_compute.cu [hunk 4]
```diff
@@ -30,10 +30,26 @@ struct GRUUnitFunctor<platform::CUDADeviceContext, T> {
...
+                                      tiled_size><<<grid, threads, 0, stream>>>(
+          value.gate_value, value.prev_out_value, value.gate_weight,
+          value.reset_output_value, frame_size, active_gate);
+
+      frame_blocks = (frame_size + tiled_size - 1) / tiled_size;
+      threads = dim3(tiled_size, 1);
```
Comment (tensor-tang) [inline paddle/fluid/operators/math/gru_compute.cu]: threads seems not changed

File: paddle/fluid/operators/math/detail/gru_gpu_kernel.h [hunk 5]
```diff
@@ -101,6 +101,114 @@ __global__ void KeGruForwardFinalOutput(OpFinalOutput op_final_output,
...
+      }
+      reset_output[COL - frame_size] = c0 * htp_0;
+    } else if (COL < frame_size) {
+      gate_value[COL] = c0;
+    }
+  }
+}
+
+template <class T, int Tiled_size>
+__global__ void KeFastCollectiveGruOut(T *gate_weight, T *prev_out_value,
```
Comment (tensor-tang) [inline paddle/fluid/operators/math/detail/gru_gpu_kernel.h:167]: same here.

File: paddle/fluid/operators/math/gru_compute.cu (hunk 1 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/math/gru_compute.cu]: ok

File: paddle/fluid/operators/math/gru_compute.cu (hunk 2 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/math/gru_compute.cu:38]: means gpu kernel collect data into register

File: paddle/fluid/operators/math/gru_compute.cu (hunk 4 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/math/gru_compute.cu]: ok

File: paddle/fluid/operators/math/detail/gru_gpu_kernel.h (hunk 5 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/math/detail/gru_gpu_kernel.h:167]: done

File: paddle/fluid/operators/math/detail/gru_gpu_kernel.h (hunk 3 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/math/detail/gru_gpu_kernel.h:108]: done

**Review by tensor-tang — APPROVED** [review]
LGTM

**Pull request merged**